File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode dummy;
        ListNode* builder = &dummy;

        while (list1 && list2) {
            if (list1->val < list2->val) {
                builder->next = list1;
                list1 = list1->next;
            }
            else {
                builder->next = list2;
                list2 = list2->next;
            }
            builder = builder->next;
        }
        if (list1) {
            builder->next = list1;
        }
        if (list2) {
            builder->next = list2;
        }

        return dummy.next;
    }
};

File: Data Structures & Algorithms/reorder-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        First Attempt:
            split the linked list in half
            reverse the second half of the linked list
            insert the reversed half
        """

        one = head
        size = 0
        while one:
            size += 1
            one = one.next
        
        ## isolate half of the linked list
        one = head
        two = head
        skipCount = math.ceil(size / 2)
        while skipCount > 0:
            two = two.next
            skipCount -= 1
        
        ## reverse second half of the linked list
        prev = None
        curr = two
        while curr is not None:
            nxt = curr.next
            curr.next = prev
            prev = curr
            curr = nxt
        two = prev

        ## merge reversed second half into first half
        ## either size(one) == size(two) or size(one) = size(two) + 1
        while one is not None and two is not None:
            one_next = one.next
            one.next = two
            two_next = two.next
            two.next = one_next
            one = one_next
            two = two_next
        one.next = None